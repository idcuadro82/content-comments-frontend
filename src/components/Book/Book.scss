@import "src/styles/colors";
@import "src/styles/constants";

@function headings($from: 1, $to: 6) {
  @if $from == $to {
    @return "h#{$from}";
  } @else {
    @return "h#{$from}," + headings($from + 1, $to);
  }
}

.document-block-item {
  &-content {
    #{headings(1,6)} {
      font-weight: bold;
      line-height: $text-line-height;
      margin-bottom: 0.5rem;
    }

    h1 {
      font-size: 2.5rem;
    }

    h2 {
      font-size: 2rem;
    }

    h3 {
      font-size: 1.5rem;
      margin-bottom: 1rem;
    }

    .block-type-bullet {
      &-text {
        margin-bottom: 0.75rem;
        padding-left: 1.5rem;
        position: relative;

        p {
          font-size: 1rem;
          line-height: $text-line-height;
        }

        &::before {
          content: "â€¢";
          font-size: 1rem;
          left: 0.75rem;
          position: absolute;
          top: 0;
        }
      }
    }

    .block-type-divider {
      border-bottom: 1px solid rgba(255, 255, 255, 0.3);
      border-top: 1px solid rgba(0, 0, 0, 0.1);
      height: 0;
      margin: 1.5rem 0;
    }
  }
}

.color-2 {
  .document-block-item {
    &-content {
      #{headings(1,6)} {
        color: $red-wine;
      }
    }
  }
}

.color-5 {
  .document-block-item {
    &-content {
      #{headings(1,6)} {
        color: $blue;
      }
    }
  }
}

.two-colums {
  display: grid;
  grid-template-columns: auto auto;
  column-gap: 40px;
}
