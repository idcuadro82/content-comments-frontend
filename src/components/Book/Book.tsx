import React, { FunctionComponent } from 'react';
import PropTypes from 'prop-types';
import PageWrapperView from '../PageWrapperView/PageWrapperView';

type BooksProps = {
  volume: string;
};

const Book: FunctionComponent<BooksProps> = ({ volume }): any => {
  return <PageWrapperView>Book {volume}</PageWrapperView>;
};

Book.propTypes = {
  volume: PropTypes.string.isRequired
};

export default Book;
