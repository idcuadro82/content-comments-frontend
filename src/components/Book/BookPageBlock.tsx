import React, { FC, useRef, useState, useEffect } from 'react'
import PropTypes from 'prop-types';
import BookPageElementModel from '../../models/BookPageContent/BookPageElementModel';
import { isBlockExcludedComment } from '../../utils/CommentBlockTypeValidator';

interface BookPageBlockProps {
  blockData?: BookPageElementModel;
  children: React.ReactNode;
}

const BookPageBlock: FC<BookPageBlockProps> = ({ blockData, children }) => {
  let [isHover, setIsHover] = useState<boolean>(false);
  let elementRef = useRef<HTMLHeadingElement>(null);

  useEffect(() => {
    elementRef.current?.addEventListener('mouseover', e => {
      e.preventDefault();
      setIsHover(true);
    })

    elementRef.current?.addEventListener('mouseleave', e => {
      e.preventDefault();
      setIsHover(false);
    })

    return () => {
      elementRef.current?.removeEventListener('mouseover', () => { });
      elementRef.current?.removeEventListener('mouseleave', () => { });
    }
  }, []);

  const validCommentBlock = () => blockData && !isBlockExcludedComment(blockData.blockType);
  const showCommentButton = () => validCommentBlock() && isHover;
  const resolveComponentClass = (): string => {
    let baseClass = 'document-block-item';
    if (blockData?.themeColor) {
      baseClass += ` ${blockData.themeColor}`
    }
    if (validCommentBlock()) {
      baseClass += ' comment-available';
    }
    return `document-block-item ${baseClass}`;
  }

  const contentProp = (Array.isArray(blockData?.content)) ? '' : blockData?.content;
  const dataContainer = Array.isArray(blockData?.content) || null;

  return (
    <div
      key={blockData?.id}
      className={resolveComponentClass()}
      data-uuid={blockData?.id}
      data-position={blockData?.position}
      data-block-type={blockData?.blockType}
      data-text={contentProp}
      data-container={dataContainer}
      ref={elementRef}>
      {
        showCommentButton() && <span className="document-block-comment" />
      }
      <div className="document-block-item-container">
        <div className="document-block-item-content">
          {children}
        </div>
      </div>
    </div>
  );
}

BookPageBlock.propTypes = {
  blockData: PropTypes.instanceOf(BookPageElementModel),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

export default BookPageBlock;
