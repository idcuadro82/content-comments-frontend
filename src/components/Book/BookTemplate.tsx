import React, { FC } from 'react';
import PropTypes from 'prop-types';
import BookPageElementModel from '../../models/BookPageContent/BookPageElementModel';
import PageElementTypeMap, { BlockTypes } from '../../models/utils/PageElementTypesMap';
import parse from 'html-react-parser';
import './Book.scss';

type BookTemplateProps = {
  content: String | BookPageElementModel[];
};

const BookTemplate: FC<BookTemplateProps> = ({ content }) => {
    
  if ((Array.isArray(content))) {
    return (
      <>
        {
          content.map((block) => {
            const contentProp = (Array.isArray(block.content)) ? '' : block.content;
            const dataContainer = Array.isArray(block.content) || null;
            return (
              <div
                key={block.id}
                className="document-block-item"
                data-uuid={block.id}
                data-position={block.position}
                data-block-type={block.blockType}
                data-text={contentProp}
                data-container={dataContainer}
              >
                <div className="document-block-item-container">
                  <div className="document-block-item-content">
                    {renderBlockContent(block)}
                  </div>
                </div>
              </div>
            );
          })
        }
      </>
    );
  }

  return (
    <div className="document-block-text">
      {content}
    </div>
  );
}

const renderBlockContent = ({
  content,
  blockType,
  themeColor
}: BookPageElementModel): JSX.Element | null => {
  switch (blockType) {

    case PageElementTypeMap.get(BlockTypes.BORDERED_CONTAINER):
      return (
        <div className={`block-type-container bordered-container ${themeColor}`}>
          <BookTemplate content={content}></BookTemplate>
        </div>
      );

    case PageElementTypeMap.get(BlockTypes.BULLET):
      return (
        <div className="block-type-bullet">
          <div className="block-type-bullet-text">
            <p contentEditable={false}>
              {content}
            </p>
          </div>
        </div>
      );

    case PageElementTypeMap.get(BlockTypes.DIVIDER):
      return (
        <hr className="block-type-divider" />
      );

    case PageElementTypeMap.get(BlockTypes.IMAGE):
      return (
        <img src={content as string} alt="" />
      );

    case PageElementTypeMap.get(BlockTypes.TEXT):
      return (
        <p contentEditable={false}>
          {parse(content as string)}
        </p>
      );

    case PageElementTypeMap.get(BlockTypes.TITLE):
      return (
        <h1 contentEditable={false}>
          {parse(content as string)}
        </h1>
      );

    case PageElementTypeMap.get(BlockTypes.TITLE_MEDIUM):
      return (
        <h2 contentEditable={false}>
          {parse(content as string)}
        </h2>
      );

    case PageElementTypeMap.get(BlockTypes.TITLE_SMALL):
      return (
        <h3 contentEditable={false}>
          {parse(content as string)}
        </h3>
      );
    default:
      return null;
  }
}

BookTemplate.propTypes = {
  content: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.instanceOf(BookPageElementModel).isRequired
    ),
    PropTypes.string
  ]).isRequired
}

export default BookTemplate;
