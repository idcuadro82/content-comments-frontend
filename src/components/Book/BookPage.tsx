import React, { FC } from 'react';
import PropTypes from 'prop-types';
import BookPageElementModel from '../../models/BookPageContent/BookPageElementModel';
import PageElementTypeMap, { BlockTypes } from '../../models/utils/PageElementTypesMap';
import parse from 'html-react-parser';
import './Book.scss';
import BookPageBlock from './BookPageBlock';

type BookPageProps = {
  content: String | BookPageElementModel[];
};

const BookPage: FC<BookPageProps> = ({ content }) => {
  if ((Array.isArray(content))) {
    return (
      <>
        {
          content.map((block) => {
            return (
              <BookPageBlock key={block.id} blockData={block}>
                {renderBlockContent(block)}
              </BookPageBlock>
            );
          })
        }
      </>
    );
  }

  return (
    <BookPageBlock>
      {content}
    </BookPageBlock>
  );
}

const renderBlockContent = ({
  content,
  blockType,
  themeColor
}: BookPageElementModel): JSX.Element | null => {
  switch (blockType) {

    case PageElementTypeMap.get(BlockTypes.BORDERED_CONTAINER):
      return (
        <div className={`block-type-container bordered-container ${themeColor}`}>
          <BookPage content={content} />
        </div>
      );

    case PageElementTypeMap.get(BlockTypes.BULLET):
      return (
        <div className="block-type-bullet">
          <div className="block-type-bullet-text">
            <p contentEditable={false}>
              {content}
            </p>
          </div>
        </div>
      );

    case PageElementTypeMap.get(BlockTypes.DIVIDER):
      return (
        <hr className="block-type-divider" />
      );

    case PageElementTypeMap.get(BlockTypes.IMAGE):
      return (
        <img src={content as string} alt="" />
      );

    case PageElementTypeMap.get(BlockTypes.TEXT):
      return (
        <p contentEditable={false}>
          {parse(content as string)}
        </p>
      );

    case PageElementTypeMap.get(BlockTypes.TITLE):
      return (
        <h1 contentEditable={false}>
          {parse(content as string)}
        </h1>
      );

    case PageElementTypeMap.get(BlockTypes.TITLE_MEDIUM):
      return (
        <h2 contentEditable={false}>
          {parse(content as string)}
        </h2>
      );

    case PageElementTypeMap.get(BlockTypes.TITLE_SMALL):
      return (
        <h3 contentEditable={false}>
          {parse(content as string)}
        </h3>
      );
    default:
      return null;
  }
}

BookPage.propTypes = {
  content: PropTypes.oneOfType([
    PropTypes.arrayOf(
      PropTypes.instanceOf(BookPageElementModel).isRequired
    ),
    PropTypes.string
  ]).isRequired
}

export default BookPage;
