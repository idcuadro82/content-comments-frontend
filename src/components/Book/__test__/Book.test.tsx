import React from 'react';
import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Book from '../Book';
import { mockGetBookContentResponse } from './GetBookContentResponse.mock';

jest.mock('../../../services/BookService.ts', () => {
  return {
    getBookContent: jest.fn().mockImplementation(() => {
      return Promise.resolve(mockGetBookContentResponse);
    })
  };
});

describe('<Book>', () => {
  it('should be equal to snapshot', async () => {
    let wrapper;
    await act(async () => {
      wrapper = render(<Book idBook="4b900a74-e2d9-4837-b9a4-9e828752716e" />);
    });
    expect(wrapper.container).toMatchSnapshot();
  });
});
