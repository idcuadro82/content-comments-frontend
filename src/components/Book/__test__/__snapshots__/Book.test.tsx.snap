// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Book> should be equal to snapshot 1`] = `
<div>
  <div
    class="document"
  >
    <div
      class="document-content"
    >
      <div
        class="document-block-list"
      >
        <div
          class="document-block-text"
        />
      </div>
    </div>
  </div>
</div>
`;
